Docs/ur sim guide (#316)
aica-technology/api: docs/docs/examples/guides/ur-sim-guide.md

---
sidebar_position: 5
title: URSim Guide
---

import Tabs from '@theme/Tabs';
import TabItem from '@theme/TabItem';

import ursimActivate from './assets/ursim-activate.gif'
import ursimHardwareParams from './assets/ursimHarwareParams.png'

# URSim guide

> This guide aims to support users with running URSim and connecting it to AICA Studio. URSim is a useful simulation

Author: Good, will rephrase this.

functionality that is not present includes:

- Emergency stop button
- Input IO state
- Self or surrounding object collisions
- Force mode

For more information, please refer to the [installation page](https://www.universal-robots.com/download/software-ur-series/simulator-linux/offline-simulator-ur-series-e-series-ur-sim-for-linux-5220/). 

TBD: Should we keep the following note?

:::note
URSim is made for Linux. For other operating systems, a virtual machine is needed. For more information, check the 
[official installation instructions](https://www.universal-robots.com/download/software-ur-series/simulator-non-linux/offline-simulator-ur-series-e-series-ur-sim-for-non-linux-5220/). 
:::

## Installation

Executing the following instructions runs URSim in a Docker container.

1. Clone the `simulators` repository provided by AICA. 

```bash
git clone https://github.com/aica-technology/simulators.git
```

TBD: must have Docker installed. Should we add something about that?
Probably people reading this have already done that. 

2. Navigate to the `ursim` folder inside this repository. 

3. Run the `run.sh` script.

```bash

> ./run.sh

Reviewer: Here we might need to specify a few more things in the script (this is not the correct line to introduce this, I'm just typing my comment here)

```suggestion
./run.sh -v 5.16.1 -m ur5e
```

v5.2x is still good for all our remote things, but is technically incompatible with some UR scripts from earlier versions. Just for the sake of being sure I think we should somehow mention this (perhaps in a note block).

We don't have to explicitly say 5.16.1, but we just need to mention the incompatibilities between 5.1x and 5.2x

fyi @domire8

Reviewer 2: really? what kind of incompatibilities?

Reviewer: move commands are now using optimove and for whatever reason this seems to create some kind of incompatibility where applications built in newer version are not backwards compatible (never looked too much in depth, but that's the jist of it)

Reviewer 2: loading .urp files you mean? Does UR have a page where this is mentioned? Otherwise I'd rather not go into explaining the incompatibilities but rather just saying that "you can use the same version as on your robot by providing the argument"

To exit, press CTRL+C
```

## Accessing and configuring the simulated robot

Follow the terminal link in a browser to access the simulated robot.

1. In the window that appears, select **Connect**. 
2. After the teaching pendant interface loads up, navigate to the settings page by clicking the burger icon 
in the top left corner of the screen.
3. Click on the **System** tab, then select the **Remote Control** tab.
4. Click **Enable** and then **Exit** at the bottom left of the screen.
5. A touchpad icon labeled **Local** should appear in the top right corner of the simulator. Click on this 
icon and then select **Remote Control**.
6. Turn on the robot by pressing the red button located in the bottom left corner of the screen. Click **ON** 
followed by **START** to activate, then click **Exit**.

<div class="text--center">
  <img src={ursimActivate} alt="Activating the simulator" />
</div>

7. The simulator is now ready to interface with an AICA application.

:::tip
Setting the robot to **Remote Control** mode is not strictly required for the simulator.
It is possible to switch to **Local Control**, where the **Installation** tab can be used to visualize 

> the robot's motions. But keep in mind that Remote Control is always required on the real robot. 

Reviewer: agree with @domire8  and this is a bit related to my comment on running a specific URSim version, because we might have UR scripts that pass the control back n' forth, and then go to the real robot and realize some of the commands are incompatible. (this is primarily due to URs OptiMove that came in these versions)

> the robot's motions. But keep in mind that Remote Control is always required on the real robot. 

Author: Yes yes, I remember, this is I have the note right below. I will continue the conversation in the other comment.

:::

> TBD: Probably not always required, since there can be exchange of control. Should we elaborate?

Author: > Yes we should elaborate. I would like these steps to just enable remote control but not to say that the switch from local mode to remote is required. So steps 1 to 4, step 6 and that's it.

I guess you mean to include the steps to activate the robot (steps 1-4 and 6), not the remote control,

> Then I would add new section here that compares local and remote control. I can help with that.

which is to move in a new section comparing the two modes.

---
sidebar_position: 8

> title: URSim Guide

Reviewer 2: ```suggestion
title: URSim guide
```

---

> sidebar_position: 5

Reviewer 2: I dont mind a lot about the order, we may have to think about that in a different  PR. for now, as long as there are no duplicates its good

# URSim guide

URSim is a software-in-the-loop simulation tool that allows the behavior of the real Universal Robots controller to be
emulated on a computer. This environment can be utilized to become familiar with the robot interface, as well as to
create and run programs. Please note that compared to the real robot, there are some limitations to the simulator.
Especially the force mode will be limited in use, since there is no physics engine to simulate valid foce values from
the sensor. Other functionality that is not present includes:

- Emergency stop button
- Input IO state
- Self or surrounding object collisions

For more information, please refer to the
[installation page](https://www.universal-robots.com/download/software-ur-series/simulator-linux/offline-simulator-ur-series-e-series-ur-sim-for-linux-5220/).

This guide aims to support users with running URSim and connecting it to AICA Studio. This is a great way to validate
application operation and safe execution before moving to the actual hardware.

## Installation

Executing the following instructions runs URSim in a Docker container. This guide assumes that users have already

> installed and configured Docker, as it is also required for AICA Studio. For more information you can check the

Author: I can change the phrasing if there is an objection but I feel they are not contradictive. We assume that they are done with Docker installation already, and in case they are not, we provide the installation link without diving in too many details here.

[installation and launch instructions](../../getting-started/installation/installation-and-launch.md) on the Getting
started page of our documentation.

Executing the following commands runs URSim in a Docker container:

```bash
git clone https://github.com/aica-technology/simulators.git
cd simulators
./run.sh
```

The last command installs and runs the URSim Docker container, simulating the UR5e robot by default. In case users want
to simulate a different robot model, the `-m` argument can be used. 

```bash
./run.sh -m ur10e
```

Another useful argument is `-v`, which defines the URSim version to be used. It is important to make sure that the URSim
version matches the Polyscope version in the actual hardware to avoid incompatibilities. Please also note that there
might be incompatibilites between 5.1x and 5.2x, with applications built in newer versions not being backwards
compatible.

```bash

> ./run.sh -v 5.16.1

Reviewer: yup, works, it's also the correct version, I don't think 5.16 is there but I haven't verified

> ./run.sh -v 5.16.1

Reviewer: we can probably link this though: https://hub.docker.com/r/universalrobots/ursim_e-series/tags

> ./run.sh -v 5.16.1

Reviewer 2: Yeah thats why i was confused

---
sidebar_position: 7
title: URSim Guide
---

import Tabs from '@theme/Tabs';
import TabItem from '@theme/TabItem';

import ursimActivate from './assets/ursim-activate.gif'
import ursimHardwareParams from './assets/ursimHarwareParams.png'

# URSim guide

URSim is a software-in-the-loop simulation tool that allows to emulate the behavior of the real Universal Robots
controller on a computer. This environment can be utilized to become familiar with the robot interface, as well as to

> create and run programs. Please note that compared to the real robot, there are some limitations to the simulator.

Reviewer 2: ```suggestion
create and run programs. Note that compared to the real robot, there are some limitations to the simulator.
```

Especially the force mode will be limited in use, since there is no physics engine to simulate valid force values from
the sensor. Other functionality that is not present includes:

- Emergency stop button
- Input IO state
- Self or surrounding object collisions

> For more information, please refer to the
> [installation page](https://www.universal-robots.com/download/software-ur-series/simulator-linux/offline-simulator-ur-series-e-series-ur-sim-for-linux-5220/).

Reviewer 2: ```suggestion
For more information, please refer to URs
[download page](https://www.universal-robots.com/download/software-ur-series/simulator-linux/offline-simulator-ur-series-e-series-ur-sim-for-linux-5220/).
```

Reviewer 3: "UR's download page" or "the UR download page"
